crypto/asn1/a_sign.c

```diff
@@ -144,6 +144,22 @@ int ASN1_item_sign_ex(const ASN1_ITEM *it, X509_ALGOR *algor1,
     return rv;
 }
 
+static int replace_algor_contents_from_DER(X509_ALGOR *algor, const unsigned char *aid, size_t len)
+{
+    /* as a workaround for d2i_*() freeing its first argument, using NULL instead: */
+    X509_ALGOR *alg = d2i_X509_ALGOR(NULL, &aid, len);
+
+    if (alg == NULL) {
+        ERR_raise(ERR_LIB_ASN1, ASN1_R_DECODE_ERROR);
+        return 0;
+    }
+    ASN1_OBJECT_free(algor->algorithm);
+    ASN1_TYPE_free(algor->parameter);
+    *algor = *alg;
+    OPENSSL_free(alg);
```

Comment:
Wouldn't it be better to finally fix `X509_ALGOR_copy()` rather than reimplementing part of it here? See #22705 and #22767.

Reply:
@DDvO please consider using X509_ALGOR_copy().